relays/ethereum/src/ethereum_client.rs

```diff
@@ -150,6 +214,172 @@ async fn transaction_receipt(client: Client, hash: H256) -> (Client, Result<Rece
+	};
+	let hash = match SubstrateHash::decode(&mut &raw_hash[..]) {
+		Ok(hash) => hash,
+		Err(error) => return (client, Err(Error::ResponseParseFailed(format!("{}", error)))),
+	};
+
+	if number != number.low_u32().into() {
+		return (client, Err(Error::InvalidSubstrateBlockNumber));
+	}
+
+	(client, Ok(HeaderId(number.low_u32(), hash)))
+}
+
+/// Returns true if Substrate header is known to Ethereum node.
+pub async fn substrate_header_known(
+	client: Client,
+	contract_address: Address,
+	id: SubstrateHeaderId,
+) -> (Client, Result<(SubstrateHeaderId, bool), Error>) {
+	let (encoded_call, call_decoder) = bridge_contract::functions::is_known_header::call(id.1);
+	let call_request = bail_on_arg_error!(
+		to_value(CallRequest {
+			to: Some(contract_address),
+			data: Some(encoded_call.into()),
+		})
+		.map_err(|e| Error::RequestSerialization(e)),
+		client
+	);
+	let (client, call_result) =
+		bail_on_error!(call_rpc::<Bytes>(client, "eth_call", Params::Array(vec![call_request]),).await);
+	match call_decoder.decode(&call_result.0) {
+		Ok(is_known_block) => (client, Ok((id, is_known_block))),
+		Err(error) => (client, Err(Error::ResponseParseFailed(format!("{}", error)))),
+	}
+}
+
+/// Submits Substrate headers to Ethereum contract.
+pub async fn submit_substrate_headers(
+	client: Client,
+	params: EthereumSigningParams,
+	contract_address: Address,
+	headers: Vec<QueuedSubstrateHeader>,
+) -> (Client, Result<(Vec<H256>, Vec<SubstrateHeaderId>), Error>) {
+	let (mut client, mut nonce) =
+		bail_on_error!(account_nonce(client, params.signer.address().as_fixed_bytes().into()).await);
+
+	let mut tx_hashes = Vec::with_capacity(headers.len());
+	let ids = headers.iter().map(|header| header.id()).collect();
+	for header in headers {
+		let (ret_client, tx_hash) = bail_on_error!(
+			submit_ethereum_transaction(
+				client,
+				&params,
+				Some(contract_address),
+				Some(nonce),
+				true, // we may need slightly more gas because other actors could change contract state
+				bridge_contract::functions::import_header::encode_input(header.extract().0.encode(),),
+			)
+			.await
+		);
+
+		nonce += 1.into();
+		tx_hashes.push(tx_hash);
+		client = ret_client;
+	}
+
+	(client, Ok((tx_hashes, ids)))
+}
+
+/// Deploy bridge contract.
+pub async fn deploy_bridge_contract(
+	client: Client,
+	params: &EthereumSigningParams,
+	contract_code: Vec<u8>,
+	initial_header: Vec<u8>,
+	initial_set_id: u64,
+	initial_authorities: Vec<u8>,
+) -> (Client, Result<H256, Error>) {
+	submit_ethereum_transaction(
+		client,
+		params,
+		None,
+		None,
+		false,
+		bridge_contract::constructor(contract_code, initial_header, initial_set_id, initial_authorities),
+	)
+	.await
+}
+
+/// Submit ethereum transaction.
+async fn submit_ethereum_transaction(
+	client: Client,
+	params: &EthereumSigningParams,
+	contract_address: Option<Address>,
+	nonce: Option<U256>,
+	double_gas: bool,
```

Comment:
How would I know when this should be used?

Reply:
Not sure, what do you mean by this question? :) The direct answer is when you want to provide `estimated_gas * 2` instead of `estimated_gas`.

Reply:
Probably [this comment](https://github.com/paritytech/parity-bridges-common/pull/65/files#diff-440221e3b161f2af0600683d4f2ab86aR290) could be helpful?

Reply:
My question was more: how would I know in what conditions I need to double the gas? That comment is helpful, but in my mind only covers one case

Reply:
That comments is the only place where `double_gas` is `true`. So it covers one case, but it is the only case :) So the answer is - when you expect that the gas requirements may change.

Just looking at the contract code again, and I've realized that the loop has (that has been in `importHeader` initially) been removed now && header import is now contextless. I'll leave `double_gas`, though, because it'll be required for next PR.